### PR TITLE
Abstract around ZCAPs for capability subsystem

### PR DESCRIPTION
**Pull request by chunningham**
Orients the HTTP API and capability subsystem around the `Delegation`, `Invocation` and `Revocation` types, which are wrappers around ZCAP invocations and delegations.

- They are only buildable via `TryFrom` and `Deserialise`, which will check content invariants related to Kepler functionality.
- When stored in the capability IPLD graph, they are transliterated into CBOR-LD.
- Revocations are a subset of Invocations
- Currently, Orbit peer delegation can only be done by a direct orbit controller
- peer delegation/orbit creation has been fixed to try and use the named peer-delegate from the delegation
- There are still bugs, mostly around inconsistent usage of the `id` field for indexing within the graph storage and similar issues

### CODE REVIEW

File: src/zcap.rs [hunk 1]
```diff
@@ -1,215 +1,448 @@
...
     let inv_str = r#"{"@context":["https://w3id.org/security/v2",{"capabilityAction":{"@id":"sec:capabilityAction","@type":"@json"}}],"id":"uuid:8097ab5c-ebd6-4924-b659-5f8009429e4d","invocationTarget":"kepler:pkh:eip155:1:0x3401fBE360502F420D5c27CB8AED88E86cc4a726://default/ipfs/#list","proof":{"type":"Ed25519Signature2018","proofPurpose":"capabilityInvocation","verificationMethod":"did:key:z6MkuMN5NfBrN6YbGjzsc5ekSQBVGut3Q6inc8aEtY2AoHZj#z6MkuMN5NfBrN6YbGjzsc5ekSQBVGut3Q6inc8aEtY2AoHZj","created":"2022-03-21T13:59:14.455Z","jws":"eyJhbGciOiJFZERTQSIsImNyaXQiOlsiYjY0Il0sImI2NCI6ZmFsc2V9..ybqGJAhCtAPE97cZTLLvX5f5IzJtZLaCmrYAGosckwt9MT5A-ZRQfcZsdwrDUGND5lSTAIAvxWjCOvtMA1RVCw","capability":"kepler:pkh:eip155:1:0x3401fBE360502F420D5c27CB8AED88E86cc4a726://default"}}"#;
-    let inv: KeplerInvocation = serde_json::from_str(inv_str)?;
-    let res = inv.verify_signature(None, DID_METHODS.to_resolver()).await;
+    let inv: Invocation = serde_json::from_str(inv_str)?;
+    let res = inv
+        .0
+        .verify_signature(None, DID_METHODS.to_resolver())
+        .await;
     assert!(res.errors.is_empty());
     Ok(())
```
Comment (clehner) [inline src/zcap.rs:447]: Test failure in `zcap::basic`:
```
---- zcap::basic stdout ----
Error: Missing Invoker
thread 'zcap::basic' panicked at 'assertion failed: `(left == right)`
  left: `1`,
 right: `0`: the test returned a termination value with a non-zero status code (1) which indicates a failure', /rustc/7737e0b5c4103216d6fd8cf941b7ab9bdbaace7c/library/test/src/lib.rs:187:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

File: src/auth.rs [hunk 2]
```diff
@@ -104,74 +94,74 @@ impl<'l> FromRequest<'l> for DelegateAuthWrapper {
...
-            Err(e) => return unauthorized(e),
-            Ok(d) => d,
+                match create_orbit(token.resource().orbit(), &config, &[], relay, kp).await {
+                    Ok(Some(orbit)) => orbit,
+                    Ok(None) => return conflict(anyhow!("Orbit already exists")),
+                    Err(e) => return internal_server_error(e),
+                }
+            }
+            (_, Ok(None)) => return not_found(anyhow!("No Orbit found")),
+            (_, Ok(Some(o))) => o,
```
[the remark this answers is not in the document]
Comment (chunningham) [reply src/auth.rs:145]: adding a `host` delegation when the orbit already exists should be perfectly legal IMO, It just shouldn't to the orbit creation process again

File: src/auth.rs [hunk 3]
```diff
@@ -104,74 +94,74 @@ impl<'l> FromRequest<'l> for DelegateAuthWrapper {
...
+                    Err(_) => {
+                        return internal_server_error(anyhow!("could not retrieve open key set"))
+                    }
+                };
 
-        let delegation = match (orbit.capabilities.root.clone(), token.delegation).try_into() {
-            Err(e) => return unauthorized(e),
-            Ok(d) => d,
+                match create_orbit(token.resource().orbit(), &config, &[], relay, kp).await {
+                    Ok(Some(orbit)) => orbit,
```
[the remark this answers is not in the document]
Comment (chunningham) [reply src/auth.rs:139]: oh, yes!

File: src/auth.rs [hunk 4]
```diff
@@ -104,74 +94,74 @@ impl<'l> FromRequest<'l> for DelegateAuthWrapper {
...
+            (true, Ok(None)) => {
+                let keys = match req
+                    .rocket()
+                    .state::<RwLock<HashMap<PeerId, Ed25519Keypair>>>()
+                {
+                    Some(k) => k,
+                    _ => return internal_server_error(anyhow!("could not retrieve open key set")),
+                };
+
+                if let Err(e) = token.verify(None).await {
```
[the remark this answers is not in the document]
Comment (chunningham) [reply src/auth.rs:119]: you're probably right, will add

**Review by cobward — APPROVED** [review]

**Pull request merged**